### ISSUE TITLE
Generating longer videos at test time

### ISSUE THREAD

**Issue by mrkulk** [lucidrains/video-diffusion-pytorch#4]
Thank you for quickly implementing this model @lucidrains ! Maybe you already have or are planning to do this -- "To manage the computational requirements of training our models, we only train on a small subset of say 16 frames at a time. However, at test time we can generate longer videos by extending our samples." (sec 3.1). Currently the sample() function is fixed length. Wanted to check in with you before taking a stab

**Comment by lucidrains** [comment]
You are welcome and no immediate plans yet for that portion. If you get to it first, do submit a PR :)

**Comment by lucidrains** [comment]
@mrkulk still planning on giving it a stab?

**Comment by mrkulk** [comment]
been planning to but came across a big issue in the stack trace before it.
I am unable to get really good temporally coherence in the basic prediction
task (gradient method is supposed to fix this but still getting choppy
predictions)

**Comment by lucidrains** [comment]
@mrkulk ahh got it, ok, i'll take a look at the gradient method tomorrow

**Comment by mrkulk** [comment]
@lucidrains still training (moving mnist) and it might reach temporal coherence after more training but this is where it's at after 40k steps. 
![W B Chart 4_28_2022, 11_45_40 PM](https://user-images.githubusercontent.com/1375144/165896626-ccb45dac-e1af-4d19-a1e4-165fbefe3592.png)
![40kiters_32x_20steps](https://user-images.githubusercontent.com/1375144/165896630-884544f6-e4bf-44f4-adb6-45a06dfbf1ee.gif)

**Comment by lucidrains** [comment]
@mrkulk cool! have you tried training it with periodic arresting of attention across time as described in the paper?

i'll have to revisit the gradient method next week

**Comment by mrkulk** [comment]
@lucidrains let me start a run now and see

**Comment by mrkulk** [comment]
much more stable but there is a temporal coherence issue. actually the problem might be deeper than inference sampling -- we would expect the predictions in just the training snippets to be consistent but it is not

![W B Chart 5_2_2022, 12_24_23 AM](https://user-images.githubusercontent.com/1375144/166199220-ef511492-533f-4f1b-a831-2a995e0e8bcc.png)

![prompt_null_42092_2abb9dc6dfed3fcd0f20](https://user-images.githubusercontent.com/1375144/166199218-e81fe3f0-7a08-4ce6-86f3-9c80af76ec1d.gif)

**Comment by lucidrains** [comment]
@mrkulk very cool! how often are you arresting the attention across time for the experiment above?

**Comment by mrkulk** [comment]
@lucidrains tried this https://github.com/lucidrains/video-diffusion-pytorch/blob/main/video_diffusion_pytorch/video_diffusion_pytorch.py#L284. Do we mean the same thing or are you referring to something else?

**Comment by lucidrains** [comment]
@mrkulk yup that's it! but what i got from the paper is that they didn't train with the arresting of attention across time exclusively? they traded off training normally vs restricting it to each frame. wasn't sure if this was done on a schedule, or alternating, or some other strategy. correct me if i'm way off base

**Comment by mrkulk** [comment]
@lucidrains it looks like they are mixing random frames at the end of a video seq -- "To implement
this joint training, we concatenate random independent image frames to the end of each video
sampled from the dataset, and we mask the attention in the temporal attention blocks to prevent
mixing information across video frames and each individual image frame. We choose these random
independent images from random videos within the same dataset;"

**Comment by mrkulk** [comment]
@lucidrains one interesting thing I noticed is that without `focus_on_the_present` turned out, it has a hard time even memorizing a single video frame. It does a good job with this turned on but its choppy even after a lot of training . Trying to see if it gets better with `focus_on_the_present=np.random.uniform(0, 1) > 0.5`  and `focus_on_the_present=True if self.global_step <= 2000 else False`
![prompt_null_4889_904d41c4ce3b1d4c1d1b](https://user-images.githubusercontent.com/1375144/166600463-9988f681-dce1-4d98-9959-55869d49cf20.gif)

.

**Comment by lucidrains** [comment]
> @lucidrains one interesting thing I noticed is that without `focus_on_the_present` turned out, it has a hard time even memorizing a single video frame. It does a good job with this turned on but its choppy even after a lot of training . Trying to see if it gets better with `focus_on_the_present=np.random.uniform(0, 1) > 0.5` and `focus_on_the_present=True if self.global_step <= 2000 else False` ![prompt_null_4889_904d41c4ce3b1d4c1d1b](https://user-images.githubusercontent.com/1375144/166600463-9988f681-dce1-4d98-9959-55869d49cf20.gif)
> 
> .

ohh nice! is that generated?

**Comment by mrkulk** [comment]
@lucidrains yes but it is a overfitting test but I suspect it will work (smooth motions won't happen or there might also be mixing of digits). I also ran a schedule (2k with focus on present and then turned it off). It didn't work as expected (although I rarely see some smooth motions).
![W B Chart 5_4_2022, 9_28_28 AM](https://user-images.githubusercontent.com/1375144/166727574-1677b3bc-a064-42b3-bd18-e586a88dc467.png)
 
It's fine around 2k when focus on present is on but then it
![062ab3f9-4abd-4715-a8d9-72095886dfe0](https://user-images.githubusercontent.com/1375144/166727746-29f0b10e-bfc0-46e2-9c3c-89cb2e81064d.gif)
![prompt_null_1873_0a8135f5c62898e9bfaa](https://user-images.githubusercontent.com/1375144/166727749-42078333-583f-43f6-9d71-d640c05acc57.gif)
 diverges -- 


@lucidrains  I am beginning to suspect its something in the Unet? Could it be attention or positional embeddings?

**Comment by lucidrains** [comment]
@mrkulk positional embedding should be fine, i'm using classic T5 relative positional bias (could even switch to a stronger one, rotary embeddings, if need be)

let me offer a way to turn off the sparse linear attention i have at every layer, and we can debug to see if that is the culprit

i have also switched from resnet blocks to the newer convnext https://arxiv.org/abs/2201.03545 , but can always bring back resnets if somehow it isn't suitable for generative work

**Comment by lucidrains** [comment]
> @lucidrains it looks like they are mixing random frames at the end of a video seq -- "To implement this joint training, we concatenate random independent image frames to the end of each video sampled from the dataset, and we mask the attention in the temporal attention blocks to prevent mixing information across video frames and each individual image frame. We choose these random independent images from random videos within the same dataset;"

i see, this must be to counter overfitting, as most videos have very similar frames. i'll think about how to build this into the trainer

**Comment by lucidrains** [comment]
@mrkulk https://github.com/lucidrains/video-diffusion-pytorch/commit/233c1d695e1a80267dac7ddd64d1d8acab17b1f6#diff-4ff1a95f5e6b9add82d0e523fd2d858ca38e67b393ea87c2ae88a8b14a0fbb1cR305 this should allow you to turn off the linear attention, in case that is causing the divergence

**Comment by lucidrains** [comment]
@mrkulk ok, i've brought back the old resnet blocks in version 0.3.1, and started a run on my own moving mnist dataset

perhaps jumping onwards to convnext wasn't the greatest idea :sweat_smile:

**Comment by mrkulk** [comment]
@lucidrains ok sounds good. will wait for your ping to do some more debugging/testing once you take a stab. btw before 0.3.1 I got errors on forward due to `SpatialLinearAttention` (goes away if you turn it off though). you may have probably already run into it

**Comment by lucidrains** [comment]
@mrkulk yup, that attention error should be fixed! here is the experiment https://wandb.ai/lucidrains/video-diffusion-redo/reports/moving-mnist-video-diffusion--VmlldzoxOTQ3OTM0?accessToken=6m0nlx9992n6pind2j3113v03tbsps52v0rtkyw4jqotpgz99ziwlx2zsh6remna also, thanks for the sponsorship! :heart:

**Comment by mrkulk** [comment]
@lucidrains awesome! the samples, even at the beginning look qualitatively different. loss seems to be steadily going down. wonder why its blobby

**Comment by lucidrains** [comment]
> @lucidrains awesome! the samples, even at the beginning look qualitatively different. loss seems to be steadily going down. wonder why its blobby

haha, i actually compared it to my previous convnext run and it looks about the same

this moving mnist dataset actually comes from https://github.com/RoshanRane/segmentation-moving-MNIST (minus the salt and pepper background noise)

i'll retry a run tonight with the new focus-on-present probabilities hyperparameter

**Comment by mrkulk** [comment]
@lucidrains i was using a different trainer but if i use your's and this below code then it seems more reasonable (only 1k iters). The moving mnist is from: `https://www.cs.toronto.edu/~nitish/unsupervised_video/`

```
from moviepy.editor import ImageSequenceClip

def moving_mnist_gif_creator():
  root = os.path.expanduser('~/datasets/mnist_test_seq.npy')
  out_root = os.path.expanduser('~/datasets/moving_mnist')
  data = np.load(root)
  for ii in range(data.shape[1]):
    clip = ImageSequenceClip(list(data[:10, ii][..., None]), fps=20)
    name = str(ii) + '.gif'
    clip.write_gif(Path(out_root) / name, fps=60)
```

![e8aa949b-b6c6-4445-8d86-58144cb2a3df](https://user-images.githubusercontent.com/1375144/166852192-9489c690-d515-4baa-ba26-0d2639c8bfa4.gif)

**Comment by lucidrains** [comment]
@mrkulk It turns out I had the wrong settings

You are right, the old resnet blocks work much better than convnext blocks, and I will likely remove them today so not to confuse researchers

![14](https://user-images.githubusercontent.com/108653/166936162-0b39aa40-ccd8-4c61-bffe-ab80ff0bc615.gif)

14k for moving mnist - it has already figured out some of the background objects, though have yet to segment the digits, but is already training way faster than a [pure attention-based solution](https://wandb.ai/lucidrains/nuwa-moving-mnist/reports/moving-mnist-nuwa--VmlldzoxNjk3MjI3?accessToken=cx03lswmr4bxj9dhrzzm5c3xebdmfq28a4dqzsoq9n89by6ppofukq7bxp19078j)

**Comment by lucidrains** [comment]
i've also added rotary positional embeddings to the temporal attention. that should definitely help no question

**Comment by lucidrains** [comment]
![sample_13000_72f8c3d673af2569cb7a](https://user-images.githubusercontent.com/108653/166938490-7f624012-020a-4971-b5f1-3eec6f00f7f6.gif)

for reference, using convnext blocks at around 13k

**Comment by mrkulk** [comment]
> 

very interesting! can give this a shot today. didn't expect convnext to not work as well. Another interesting and somewhat related point --the transframer (https://arxiv.org/pdf/2203.09494.pdf) paper used NF-ResNet block within a 3d Unet for video generation. but still need to mentally consolidate and think about these variations

**Comment by lucidrains** [comment]
![sample_19000_6697151caa3aa2553f54](https://user-images.githubusercontent.com/108653/167164421-9453d5a7-38a3-434c-95d3-f2a81d0fa3b4.gif)

@mrkulk cool! was not aware of transframer - will need to queue that up for reading

second day of training, it has yet again improved (but still has not segmented the digits yet as the attention based one has, but it is still early in training)

there are these weird flickering artifacts as well, from frame to frame, and i'll have to see if they persist once i move the training onto some remote compute that came my way

**Comment by mrkulk** [comment]
> ![sample_19000_6697151caa3aa2553f54](https://user-images.githubusercontent.com/108653/167164421-9453d5a7-38a3-434c-95d3-f2a81d0fa3b4.gif)
> 
> @mrkulk cool! was not aware of transframer - will need to queue that up for reading
> 
> second day of training, it has yet again improved (but still has not segmented the digits yet as the attention based one has, but it is still early in training)
> 
> there are these weird flickering artifacts as well, from frame to frame, and i'll have to see if they persist once i move the training onto some remote compute that came my way

it seems to have stabilized on the background and filling in higher frequency details -- quite  promising! 

how many videos are you training this on and what batch size? 

ran a test on 10k videos on the same dataset but smaller batch size (on a V100). This is without sparse linear attn. I think the trend is similar? :

https://wandb.ai/csmai/multimodal-video-generation/reports/moving-mnist-video_diffusion_model--VmlldzoxOTYwNjY0?accessToken=q26z0simmgj0zqjc61kfsixezrzvo3clidjkuujwl5ttrpivpoj3c5idj886qh4c

Seems like the digits will come towards the very end but it interesting that pure attn gets it. How many global steps are you in compared to the pure attn based one?

**Comment by lucidrains** [comment]
![sample_17000_8c11ca2aaa5f320eba55](https://user-images.githubusercontent.com/108653/167443664-9ddc26b9-8538-4b18-84f5-afcf981024df.gif) i'm about 52k steps in

it is converging much faster than the [pure attention-based one](https://wandb.ai/lucidrains/nuwa-moving-mnist/reports/moving-mnist-nuwa--VmlldzoxNjk3MjI3?accessToken=cx03lswmr4bxj9dhrzzm5c3xebdmfq28a4dqzsoq9n89by6ppofukq7bxp19078j). digits are starting to appear about now, although malformed

**Comment by lucidrains** [comment]
once this run is successful, i'll move on to conditioning (with the digit(s) on the screen)

**Comment by lucidrains** [comment]
![sample_20000_ccc016850c4582ceabd7](https://user-images.githubusercontent.com/108653/167462510-17872891-f350-45bb-84ed-52ce6345b965.gif)

54k - the flickering is the only thing that doesn't seem right - the only-attention model didn't have this problem

**Comment by mrkulk** [comment]
@lucidrains running it without sparse linear attn (wandb job i shared above) also shows some flickers but it seems much less so (maybe it is the same and its just me). it does seem like it will eventually learn the characters

**Comment by liangbingzhao** [comment]
so the focus_present_mask, is it necessary? looking forward to your guys further research

**Comment by mrkulk** [comment]
@martinriven focus_present_mask might be needed for video modeling more prominently. I tried a conditional version of this model on past frames and 0.2 seems to work well on a few datasets that were tried

**Comment by liangbingzhao** [comment]
> @martinriven focus_present_mask might be needed for video modeling more prominently. I tried a conditional version of this model on past frames and 0.2 seems to work well on a few datasets that were tried

thx, i will give it a try

**Comment by liangbingzhao** [comment]
any clue about the gradient method? how to generate a longer video?

**Comment by liangbingzhao** [comment]
https://wandb.ai/zlbf/64best&0.2probfocus/reports/Untitled-Report--VmlldzoyMzQxMjc1?accessToken=nvvkv2jgl118sxopem40jv3p11yb0llw6ivoloe9e1uihft0y1ny078b8yvxqwmx
Here is what i run through the dataset ucf101, with 0.2 focus_present_mask.  It wave a lot.And I found if I generate 64*64 video, its crest is 0.2; 128*128 video crest to 0.4; 256*256 video crest to 0.8, which is really driving me mad. Any suggestions about how to improve it? qwq

**Comment by liangbingzhao** [comment]
@lucidrains can video diffusion use the efficient unet proposed by Imagen? wonder if it could make the model better?

**Comment by mrkulk** [comment]
The loss is much smoother with arrested attention but no sign of temporal
coherence yet with 10 frame modeling after 40k steps

**Comment by farhad-abdi** [comment]
Hi, my issue was not really significant to open a thread. I just want to know how to display final sampled_videos using matplotlib, i'm permuting (channels, frames, height, width) using  `imgs = sampled_videos[0].permute(1,2,3,0)` but it's showing noise only

**Comment by yangzhao1230** [comment]
@mrkulk Hi, I wonder if you have implemented the function that generating longer videos?

**Comment by UESTC-LXD** [comment]
> mnist_test_seq
Do you use normalization on the dataset during training?

**Comment by bellonet** [comment]
I also would like this feature. Is it implemented? 
Also, the performance is amazing on my relatively small training set. Thank you!

**Comment by zzzbbt** [comment]
> @mrkulk Hi, I wonder if you have implemented the function that generating longer videos?

Did you manage to generate a longer video?
